src/grafana_service.ts

```diff
@@ -21,6 +22,82 @@ export class DatasourceUnavailable extends DataKitError {};
 
 const CHUNK_SIZE = 50000;
 
+export class TimeSeriesPoint {
+  constructor(
+    public columns: string[],
+    public values: number[]
+  ){}
+}
+
+export class TimeSeriesChunk {
+  constructor(
+    public columns: string[],
+    public values: number[]
+  ){}
+}
+
+export class DatasourceStream {
+
+  private currentChunk: TimeSeriesPoint[] = [];
+  private grafanaUrl;
+
+  constructor(
+    public metric: Metric,
+    public url: string,
+    public apiKey: string
+  ) {
+    this.grafanaUrl = getGrafanaUrl(url);
+  };
+
+  public async query(from: number, to: number): Promise<Readable> {
+
+    if(from > to) {
+      throw new BadRange(
+        `Data-kit got wrong range: from ${from} > to ${to}`,
+        this.metric.datasource.type,
+        this.url
+      );
+    }
+
+    if(from === to) {
+      console.warn(`Data-kit got from === to`);
+    }
+
+    const data = this._query(from, to);
+    const read = async (self: Readable, size: number) => {
+      if(this.currentChunk.length < size) {
+        let chunk = (await data.next()).value;
```

Comment:
it's possible that `next()` returns undefined or throw exception. It's because you can't guarantee `data` size